File: sudoq-app/sudoqmodel/src/test/java/de/sudoq/model/sudoku/sudokuTypes/StandardSudokuTypeTest.java
package de.sudoq.model.sudoku.sudokuTypes;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

import org.junit.Test;

import de.sudoq.model.sudoku.Constraint;
import de.sudoq.model.sudoku.Position;
import de.sudoq.model.sudoku.complexity.Complexity;
import de.sudoq.model.sudoku.complexity.ComplexityConstraint;

public class StandardSudokuTypeTest {

	SudokuType sst = TypeBuilder.getType(SudokuTypes.standard9x9);

	@Test
	public void test() {

		Position p = sst.getSize();
		assertNotNull("getSize returns null!", p);
		assertEquals("getX does not return 9!", 9, p.getX());
		assertEquals("getY does not return 9!", 9, p.getY());
	}

	@Test
	public void nonQuadraticBlocksTest() {
		/*SudokuType ss18 = new SST18x18();
		boolean bounds = true;
		Position p0 = Position.get(0, 0);
		Position p1 = Position.get(5, 0);
		Position p2 = Position.get(0, 2);
		Position p3 = Position.get(5, 2);

		for (Constraint c : ss18) {
			if (c.toString().contains("Block 0")) {

				boolean exists = false;
				for (Position p : c)
					if (p.equals(p0))
						exists = true;
				bounds &= exists;

				exists = false;
				for (Position p : c)
					if (p.equals(p1))
						exists = true;
				bounds &= exists;

				exists = false;
				for (Position p : c)
					if (p.equals(p2))
						exists = true;
				bounds &= exists;

				exists = false;
				for (Position p : c)
					if (p.equals(p3))
						exists = true;
				assertTrue(bounds &= exists);
			}

		}*/
	}

	public class SST18x18 extends SudokuType {
		public SST18x18() {
			//TODO somehow create this super(18);
		}

		@Override
		public SudokuTypes getEnumType() {
			return null;
		}

		@Override
		public ComplexityConstraint buildComplexityConstraint(Complexity complexity) {
			return null;
		}

		@Override
		public float getStandardAllocationFactor() {
			return 0;
		}
	}

	@Test
	public void getEnumTypeTest() {
		assertTrue(sst.getEnumType() == SudokuTypes.standard9x9);
	}

	@Test
	public void getStandartAllocationFactorTest() {
		assertTrue(sst.getStandardAllocationFactor() == 0.35f);
	}

	@Test
	public void complexityTest() {
		SudokuType type = TypeBuilder.getType(SudokuTypes.standard9x9);
		testComplexity(type, Complexity.easy,      35, 45,  400,  1200, 2);
		testComplexity(type, Complexity.medium,    27, 35, 1200,  2500, 3);
		testComplexity(type, Complexity.difficult, 22, 28, 2500,  4000, Integer.MAX_VALUE);
		testComplexity(type, Complexity.infernal,  17, 24, 4000, 25000,	Integer.MAX_VALUE);
		assertNull(type.buildComplexityConstraint(null));
	}

	private void testComplexity(SudokuType type,
								Complexity complexity,
	                            int minFields, int maxFields,
	                            int minComplexityIdentifier,
	                            int maxComplexityIdentifier,
	                            int numberOfAllowedHelpers) {
		ComplexityConstraint constraint = type.buildComplexityConstraint(complexity);
		assertEquals("expected complexity is: ", constraint.getComplexity(),              complexity);
		assertEquals("expected minComplexity:", constraint.getMinComplexityIdentifier(), minComplexityIdentifier);
		assertEquals("expected maxComplexity:", constraint.getMaxComplexityIdentifier(), maxComplexityIdentifier);
		assertEquals("expected number of allowed Helpers:", constraint.getNumberOfAllowedHelpers(),  numberOfAllowedHelpers);
	}

}

File: sudoq-app/sudoqmodel/src/test/kotlin/de/sudoq/model/sudoku/sudokuTypes/StandardSudokuTypeTest.kt
package de.sudoq.model.sudoku.sudokuTypes

import de.sudoq.model.sudoku.Position
import org.amshove.kluent.`should be`
import org.amshove.kluent.`should contain all`
import org.amshove.kluent.`should not be null`
import org.junit.jupiter.api.Test

class StandardSudokuTypeTest {
    var sst = TypeBuilder.getType(SudokuTypes.standard9x9)

    @Test
    fun test() {
        val p = sst.size
        p.`should not be null`()
        p.x.`should be`(9)
        p.y.`should be`(9)
    }

    @Test
    fun nonQuadraticBlocksTest() {
        val ss18: SudokuType = SST18x18()
        val positions = setOf(
            Position[0, 0], Position[5, 0],
            Position[0, 2], Position[5, 2]
        )
        for (c in ss18) {
            if (c.toString().contains("Block 0")) {
                c `should contain all` positions
            }
        }
    }


    @Test
    fun enumTypeTest() {
        sst.enumType `should be` SudokuTypes.standard9x9
    }
}

class SST18x18 : SudokuType(SudokuTypes.standard4x4, 9, 0f, Position[1,1], Position[1,1],
    ArrayList(), ArrayList(), ArrayList(),
    ComplexityConstraintBuilder(HashMap())) {

    override fun getStandardAllocationFactor(): Float = 0f

}
